rust/tools/authorization-logic/transparent_release_prototype/transparent_release.auth_logic

```diff
@@ -0,0 +1,276 @@
+// (transparent-release/common/common.go::Build())
+// When this runs successful, we know that the provenance file really produces 
+// a binary. So we trust this builder to say that a principal representing the 
+// binary is built by the provenance file.
+//
+// Wrapper inputs:
+// - A path to the provenance file 
+// - principal name for application
+// Internally:
+//  - the wrapper calls `LoadBuildConfigFromProvenance` on the provenance file 
+//  to generate a `BuildConfig` object (or the appropriate call sequence for 
+//  getting a BuildConfig from a provenance file is done)
+//  - The `BuildConfig` is passed to `Build()`. If this succeeds the output is 
+//  the following.
+//
+// The arguments in this example are:
+// - provenance file: "provenance_file.toml"
+// - principal name for application: "OakFunctionsLoader"
+
+"ProvenanceFileBuilder" says {
+    "Binary_OakFunctionsLoader"
+        hasProvenance("ProvenanceFile_OakFunctionsLoader").
+}
+
+//-----------------------------------------------------------------------------
+// Hash Check Wrapper (Checks hashes on files)
+//-----------------------------------------------------------------------------
+// We need a wrapper (another standalone program) that outputs a measurement
+// (sha256) of various files as authorization logic (provenance file, binary).
+// This wrapper will compute the sha256 directly (for example, using the CLI program
+// "sha256sum", or by calling 
+// [the code in the transparent release repository that checks
+// hashes](https://github.com/project-oak/transparent-release/blob/2c8f8292cab31dbacbf5f4352b51025599ae78c3/common/common.go#L316)
+// ).
+//
+// The hash check wrapper for the binary can run on the "output_path" from 
+// transparent-release/common/BuildConfig, which is where the binary is stored 
+// after it is built.
+//
+// (You could imagine extending transparent-release/common::Build() to just output 
+// the hash instead, but then there would be two principals trusted for 
+// producing hashes instead of just one, so the trust assumptions would be less 
+// clear.)
+
+// The arguments to this program are:
+// - the name of the file to be measured (hash checked)
+// - the name of the principal for the file being measured
+
+// We will invoke this wrapper twice: once for the provenance file and once
+// for the binary generated by executing the steps in the provenance file
+
+// For the provenance file, the inputs are:
+//    file name --> "provenance_file.toml"
+//    principal name --> "ProvenanceFile_OakFunctionsLoader"
+// The output looks like:
+"Sha256Wrapper" says {
+    measured_hash("ProvenanceFile_OakFunctionsLoader",
+      "6de656f11599c11548519af6b7385504ed845a4ff6600745cb3d9e174b8956af").
+}
+
+// For the binary, the inputs are:
+//    file name --> "oak_functions_loader.exe" (I don't have an example binary)
+//    principal name --> "Binary_OakFunctionsLoader"
+"Sha256Wrapper" says {
```

Comment:
Yes, it would be possible to have the endorsement file and provenance wrappers just output the hash measurements. The reason why a separate wrapper _just_ for checking hashes is used here is better separation of trust. By putting the code for checking hashes in a separate wrapper, the policy only needs to trust this wrapper for getting hashes and this trust assumption is clear in the policy code because it says `"Verifier_OakFunctionsLoader" says "Sha256Wrapper" canSay measured_hash(...)`. Because the only thing the `Verifier...` trusts this wrapper to do is check hashes, someone auditing this code base would just need to look at this wrapper and check that all it does is check hashes. If we made the provenance/endorsement generators also produce the hash, these wrappers would need to be trusted to do more things (e.g. the provenance file wrapper would be trusted both to claim that a binary was built from a provenance file and that it has some hash). If we write bigger pieces of code that are trusted to do more things we are getting a bit less value out of this language that is meant to make trust assumptions clear and narrow.

If we do it this way, I don't think we need to store the binary anywhere publicly accessible either. This verifier process using authorization logic is also meant to produce just a true/false answer to the top level query. This verifier process is also just a bazel rule / makefile / or binary that calls a bunch of other programs.
The verifier process can call [your code for building binaries from provenance files](https://github.com/project-oak/transparent-release/blob/main/build/build.go) which will make a binary on the local machine, then it can call the `sha256wrapper` code to check the hash and emit the auth logic facts. This all happens locally to the machine running the verifier process. (We would just need to pass the `sha256wrapper` the directory where the builder makes the binary which I think is `BuldConfig.OutputPath`.)

Reply:
Sounds good. Thanks for clarifying :)